File: macclean/cleaners/fonts.py
from collections import defaultdict
from pathlib import Path
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from macclean.core.utils import CleanItem, AnalysisResult, format_size, confirm

console = Console()

_FONT_DIRS = [
    Path("/Library/Fonts"),
    Path("/System/Library/Fonts"),
]

_FONT_EXTS = {".ttf", ".otf", ".ttc", ".dfont"}


def _find_duplicates(dirs: list[Path]) -> dict[str, list[Path]]:
    seen: dict[str, list[Path]] = defaultdict(list)
    for d in dirs:
        if not d.exists():
            continue
        for f in d.iterdir():
            if f.suffix.lower() in _FONT_EXTS:
                seen[f.name].append(f)
    return {name: paths for name, paths in seen.items() if len(paths) > 1}


def analyze(home: Path | None = None) -> AnalysisResult:
    home = home or Path.home()
    user_fonts = home / "Library" / "Fonts"
    dirs = [user_fonts] + _FONT_DIRS
    dupes = _find_duplicates(dirs)
    result = AnalysisResult()

    for name, paths in dupes.items():
        user_copies = [p for p in paths if str(p).startswith(str(home))]
        for path in user_copies:
            result.items.append(CleanItem(
                label=f"{name} (duplicate in {path.parent})",
                path=path,
                size_bytes=path.stat().st_size,
                removable=True,
            ))

    return result


def clean(result: AnalysisResult, dry_run: bool = False, yes: bool = False) -> None:
    if not result.items:
        console.print("[green]No duplicate fonts found.[/]")
        return

    table = Table(show_header=True, show_lines=True)
    table.add_column("Font")
    table.add_column("Size", justify="right")
    for item in result.items:
        table.add_row(item.label, format_size(item.size_bytes))

    console.print(Panel(table, title="[bold cyan]Duplicate Fonts[/]"))
    console.print(f"  Total recoverable: [bold]{format_size(result.total_bytes)}[/]")
    console.print("  [dim]Only user-installed duplicates shown (~/Library/Fonts). System fonts are not touched.[/]")

    if dry_run:
        return
    if not yes and not confirm("Remove duplicate user fonts?"):
        return

    for item in result.items:
        try:
            item.path.unlink()
            console.print(f"  [green]+[/] Removed {item.label}")
        except Exception as e:
            console.print(f"  [yellow]![/] {item.label}: {e}")


@click.command()
@click.pass_context
def cmd(ctx):
    """Detect and remove duplicate fonts in ~/Library/Fonts."""
    result = analyze()
    clean(result, dry_run=ctx.obj["dry_run"], yes=ctx.obj["yes"])

File: src/cleaners/fonts.rs
use crate::core::{AnalysisResult, CleanItem, CleanKind, Cleaner, RiskLevel};
use crate::ui;
use anyhow::Result;
use std::collections::HashMap;
use std::path::PathBuf;

pub struct FontsCleaner;

impl Cleaner for FontsCleaner {
    fn name(&self) -> &str {
        "fonts"
    }
    fn display_name(&self) -> &str {
        "Duplicate Fonts"
    }

    fn analyze(&self) -> Result<AnalysisResult> {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));
        let mut result = AnalysisResult::default();

        let font_dirs = [
            home.join("Library/Fonts"),
            PathBuf::from("/Library/Fonts"),
            PathBuf::from("/System/Library/Fonts"),
        ];

        let font_extensions = ["ttf", "otf", "ttc", "dfont"];

        // Map from lowercase filename -> list of full paths
        let mut filename_map: HashMap<String, Vec<PathBuf>> = HashMap::new();

        for dir in &font_dirs {
            if !dir.exists() {
                continue;
            }
            let entries = match std::fs::read_dir(dir) {
                Ok(e) => e,
                Err(_) => continue,
            };
            for entry in entries.filter_map(|e| e.ok()) {
                let path = entry.path();
                if !path.is_file() {
                    continue;
                }
                let ext = path
                    .extension()
                    .map(|e| e.to_string_lossy().to_lowercase())
                    .unwrap_or_default();
                if font_extensions.contains(&ext.as_str()) {
                    let name = path
                        .file_name()
                        .map(|n| n.to_string_lossy().to_lowercase())
                        .unwrap_or_default();
                    filename_map.entry(name).or_default().push(path);
                }
            }
        }

        // For duplicates, add only user-installed copies (those under ~/Library/Fonts)
        let user_font_dir = home.join("Library/Fonts");
        for paths in filename_map.values() {
            if paths.len() <= 1 {
                continue;
            }
            for path in paths {
                if path.starts_with(&user_font_dir) {
                    let parent = path
                        .parent()
                        .map(|p| p.display().to_string())
                        .unwrap_or_default();
                    let file_name = path
                        .file_name()
                        .map(|n| n.to_string_lossy().to_string())
                        .unwrap_or_default();
                    let size = path.metadata().map(|m| m.len()).unwrap_or(0);
                    result.items.push(CleanItem {
                        label: format!("{} (duplicate in {})", file_name, parent),
                        path: path.clone(),
                        size_bytes: size,
                        removable: true,
                        kind: CleanKind::Duplicate,
                        risk: RiskLevel::Medium,
                        reason: "User font has the same filename as another installed font.".into(),
                    });
                }
            }
        }

        Ok(result)
    }

    fn clean(&self, result: &AnalysisResult, dry_run: bool, yes: bool) -> Result<()> {
        if result.items.is_empty() {
            println!("No duplicate user fonts found.");
            return Ok(());
        }
        ui::print_analysis("Duplicate Fonts", &result.items);
        ui::print_warn("Only user-installed duplicates shown. System fonts will not be touched.");
        if dry_run {
            return Ok(());
        }
        if !yes && !ui::confirm("Remove duplicate user fonts?", false)? {
            return Ok(());
        }

        for (path, outcome) in crate::core::trash::trash_clean_items("fonts", &result.items) {
            match outcome {
                Ok(_) => ui::print_ok(&format!("Moved to Trash: {}", path.display())),
                Err(e) => ui::print_warn(&format!("{}: {}", path.display(), e)),
            }
        }
        Ok(())
    }
}
